File: 543.diameter-of-binary-tree/solution.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int DiameterOfBinaryTree(TreeNode root) {
        return DiameterOfBinaryTree2(root)[0];
    }

    private int[] DiameterOfBinaryTree2(TreeNode root) {
        if (root == null)
            return new int[]{0, 0};
        int[] left = DiameterOfBinaryTree2(root.left);
        int[] right = DiameterOfBinaryTree2(root.right);
        return new int[]{Math.Max(left[1] + right[1], Math.Max(left[0], right[0])), Math.Max(left[1], right[1]) + 1};
    }
}

File: 543.diameter-of-binary-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        return diameterOfBinaryTree2(root)[0];
    }

    private int[] diameterOfBinaryTree2(TreeNode root) {
        if (root == null)
            return new int[]{0, 0};
        int[] left = diameterOfBinaryTree2(root.left);
        int[] right = diameterOfBinaryTree2(root.right);
        return new int[]{Math.max(left[1] + right[1], Math.max(left[0], right[0])), Math.max(left[1], right[1]) + 1};
    }
}
